backend/src/main/java/gov/cdc/usds/simplereport/validators/FileValidator.java

```diff
@@ -0,0 +1,81 @@
+package gov.cdc.usds.simplereport.validators;
+
+import static gov.cdc.usds.simplereport.validators.CsvValidatorUtils.getIteratorForCsv;
+import static gov.cdc.usds.simplereport.validators.CsvValidatorUtils.getNextRow;
+
+import com.fasterxml.jackson.databind.MappingIterator;
+import gov.cdc.usds.simplereport.api.model.errors.CsvProcessingException;
+import gov.cdc.usds.simplereport.api.model.filerow.FileRow;
+import gov.cdc.usds.simplereport.service.model.reportstream.FeedbackMessage;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class FileValidator<T extends FileRow> {
+  private final Function<Map<String, String>, T> fileRowConstructor;
+
+  public FileValidator(Function<Map<String, String>, T> fileRowConstructor) {
+    this.fileRowConstructor = fileRowConstructor;
+  }
+
+  public List<FeedbackMessage> validate(InputStream csvStream) {
+    final MappingIterator<Map<String, String>> valueIterator = getIteratorForCsv(csvStream);
+
+    if (!valueIterator.hasNext()) {
+      throw new IllegalArgumentException("Empty or invalid CSV submitted");
+    }
+
+    var mapOfErrors = new HashMap<String, FeedbackMessage>();
+
+    while (valueIterator.hasNext()) {
+      final Map<String, String> row;
+      final var finalCurrentRow = valueIterator.getCurrentLocation().getLineNr();
+      try {
+        row = getNextRow(valueIterator);
+      } catch (CsvProcessingException ex) {
+        log.error("Unable to parse csv.", ex);
+        var feedback =
+            new FeedbackMessage(
+                CsvValidatorUtils.REPORT_SCOPE,
+                "File has the incorrect number of columns or empty rows. Please make sure all columns match the data template, and delete any empty rows.",
+                new ArrayList<>(List.of(ex.getLineNumber())));
+        mergeErrors(mapOfErrors, new ArrayList<>(List.of(feedback)));
+        continue;
+      }
+      var currentRowErrors = new ArrayList<FeedbackMessage>();
+
+      var fileRow = fileRowConstructor.apply(row);
+
+      currentRowErrors.addAll(fileRow.validateHeaders());
+      currentRowErrors.addAll(fileRow.validateIndividualValues());
+
+      currentRowErrors.forEach(
+          error -> error.setIndices(new ArrayList<>(List.of(finalCurrentRow))));
```

Comment:
Thanks for explaining that to me! I wasn't really sure what the difference was between those two, but so does that means that that even the [csv error thing](https://github.com/CDCgov/prime-simplereport/pull/4730/files/6969cdc13b6e80df3c06fc913ed9da13705a40ac#diff-2b537a592254778dd76f364d15df290173d0d8fda42a9db279bde9f26f9f911dR43-R48) should also be `ITEM_SCOPE`? If that happens no where in the code are we using `REPORT_SCOPE` so it might be easy to delete that field from `ValueOrError`.

Reply:
oh yup, you're right I think that should be ITEM_SCOPE too
I think if we had a validation that checked the header line for missing required fields (this is how RS does it) then that would be REPORT_SCOPE, but the way we have things written now it looks like we don't need that value at all

Reply:
done in [77f7cf4](https://github.com/CDCgov/prime-simplereport/pull/4730/commits/77f7cf4310ed9e54caf780fdc0524c8c4b69ad2b)/[9d118b2](https://github.com/CDCgov/prime-simplereport/pull/4730/commits/9d118b2f0511931d7ca497f6a098c5ea906d140a)